src/common/finite_element_state.hpp

```diff
@@ -0,0 +1,159 @@
+   * @param[in] order The order of the problem
+   * @param[in] pmesh The problem mesh
+   * @param[in] options The options specified, namely those relating to the dimension of the FESpace, the type of
+   * FEColl, the DOF ordering that should be used, and the name of the field
+   */
+  FiniteElementState(const int order, std::shared_ptr<mfem::ParMesh> pmesh, FESOptions&& options = FESOptions());
+
+  /**
+   * Returns the MPI communicator for the state
+   */
+  MPI_Comm comm() const { return space_.GetComm(); }
+
+  /**
+   * Returns a non-owning reference to the internal grid function
+   */
+  mfem::ParGridFunction& gridFunc() { return *gf_; }
+
+  /**
+   * Returns a non-owning reference to the internal mesh object
+   */
+  mfem::ParMesh& mesh() { return *mesh_; }
+
+  /**
+   * Returns a non-owning reference to the internal FESpace
+   */
+  mfem::ParFiniteElementSpace& space() { return space_; }
+
+  /**
+   * Returns a non-owning const reference to the internal FESpace
+   */
+  const mfem::ParFiniteElementSpace& space() const { return space_; }
+
+  /**
+   * Returns a non-owning reference to the vector of true DOFs
+   */
+  mfem::HypreParVector& trueVec() { return true_vec_; }
+
+  /**
+   * Returns the name of the FEState (field)
+   */
+  std::string name() const { return name_; }
+
+  /**
+   * Projects a coefficient (vector or scalar) onto the field
+   * @param[in] coef The coefficient to project
+   */
+  void project(const GeneralCoefficient& coef)
+  {
+    // The generic lambda parameter, auto&&, allows the component type (mfem::Coef or mfem::VecCoef)
+    // to be deduced, and the appropriate version of ProjectCoefficient is dispatched.
+    std::visit([this](auto&& concrete_coef) { gf_->ProjectCoefficient(*concrete_coef); }, coef);
+  }
+
+  /**
+   * Projects a coefficient (vector or scalar) onto the field
+   * @param[in] coef The coefficient to project
+   */
+  void project(mfem::Coefficient& coef) { gf_->ProjectCoefficient(coef); }
+
+  /**
+   * Projects a coefficient (vector or scalar) onto the field
+   * @param[in] coef The coefficient to project
+   */
+  void project(mfem::VectorCoefficient& coef) { gf_->ProjectCoefficient(coef); }
+
+  /**
+   * Initialize the true DOF vector by extracting true DOFs from the internal
+   * grid function into the internal true DOF vector
+   */
+  void initializeTrueVec() { gf_->GetTrueDofs(true_vec_); }
+
+  /**
+   * Set the internal grid function using the true DOF values
+   */
+  void distributeSharedDofs() { gf_->SetFromTrueDofs(true_vec_); }
+
+  /**
+   * Utility function for creating a tensor, e.g. mfem::HypreParVector,
+   * mfem::ParBilinearForm, etc on the FESpace encapsulated by an FEState object
+   * @return An owning pointer to a heap-allocated tensor
+   * @pre Tensor must have the constructor Tensor::Tensor(ParFiniteElementSpace*)
+   */
+  template <typename Tensor>
+  std::unique_ptr<Tensor> createTensorOnSpace()
```

Comment:
While I like the idea of this function, I'm not sure `Tensor` is appropriate since it can be used to construct explicitly nonlinear things. Maybe `createObjectOnSpace` or just `createOnSpace`? Any other ideas?

Reply:
Yes, this is kind of the same problem as `SystemSolver` (TBD?).  I was hoping for something more descriptive but since the function really only enforces the existence of a matching constructor maybe `createOnSpace` is appropriate.